File: backend/app/schemas/listing.py
"""Pydantic request/response schemas for real estate listings."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


ListingStatus = Literal["Draft", "Active", "Pending", "Sold", "Archived"]
PropertyType = Literal[
    "Single Family",
    "Condo",
    "Townhouse",
    "Multi-Family",
    "Land",
    "Commercial",
    "Farm/Ranch",
    "Manufactured Home",
    "Other",
]
HoaFeeFrequency = Literal["Monthly", "Quarterly", "Annually"]

MAX_PRICE = 10_000_000_000
MAX_SQFT = 10_000_000
MAX_ACREAGE = 100_000_000
MAX_MONEY_FIELD = 100_000_000
MAX_BEDROOMS = 100
MAX_BATHROOMS = 100
MAX_AMENITIES = 100
MAX_AMENITY_LENGTH = 100


def _normalize_state(value: str) -> str:
    """Normalize a two-letter state code."""
    normalized = value.strip().upper()
    if len(normalized) != 2 or not normalized.isalpha():
        raise ValueError("State must be a two-letter code")
    return normalized


def _validate_year_built(value: int | None) -> int | None:
    """Keep construction years within a realistic historical/future range."""
    if value is None:
        return None

    latest_year = datetime.now(timezone.utc).year + 1
    if value < 1600 or value > latest_year:
        raise ValueError(f"Year built must be between 1600 and {latest_year}")

    return value


def _normalize_amenities(values: list[str] | None) -> list[str]:
    """Trim, validate, and de-duplicate structured amenity names."""
    if not values:
        return []

    if len(values) > MAX_AMENITIES:
        raise ValueError(f"No more than {MAX_AMENITIES} amenities are allowed")

    normalized: list[str] = []
    seen: set[str] = set()

    for value in values:
        amenity = value.strip()
        if not amenity:
            continue
        if len(amenity) > MAX_AMENITY_LENGTH:
            raise ValueError(
                f"Each amenity must be {MAX_AMENITY_LENGTH} characters or fewer"
            )

        key = amenity.casefold()
        if key not in seen:
            seen.add(key)
            normalized.append(amenity)

    return normalized


class ListingBase(BaseModel):
    """Shared validated fields used when creating and reading listings."""

    title: str = Field(..., min_length=1, max_length=150)
    status: ListingStatus = "Draft"
    is_public: bool = False
    is_featured: bool = False
    hide_exact_address: bool = False

    price: int = Field(..., ge=0, le=MAX_PRICE)
    property_type: PropertyType | None = None

    address: str = Field(..., min_length=1, max_length=255)
    city: str = Field(..., min_length=1, max_length=100)
    state: str = Field(..., min_length=2, max_length=2)

    description: str | None = Field(None, max_length=20_000)
    sqft: int | None = Field(None, ge=0, le=MAX_SQFT)
    acreage: float | None = Field(None, ge=0, le=MAX_ACREAGE)
    year_built: int | None = None

    bedrooms: int = Field(..., ge=0, le=MAX_BEDROOMS)
    bathrooms: float = Field(
        ...,
        ge=0,
        le=MAX_BATHROOMS,
        multiple_of=0.5,
    )

    annual_property_taxes: float | None = Field(
        None,
        ge=0,
        le=MAX_MONEY_FIELD,
    )
    hoa_fee: float | None = Field(None, ge=0, le=MAX_MONEY_FIELD)
    hoa_fee_frequency: HoaFeeFrequency | None = None

    school_district: str | None = Field(None, max_length=150)
    amenities: list[str] = Field(default_factory=list)

    mls_number: str | None = Field(None, max_length=100)
    source_attribution: str | None = Field(None, max_length=255)
    cover_image: str | None = Field(None, max_length=2048)

    @field_validator("state")
    @classmethod
    def normalize_state(cls, value: str) -> str:
        return _normalize_state(value)

    @field_validator("year_built")
    @classmethod
    def validate_year_built(cls, value: int | None) -> int | None:
        return _validate_year_built(value)

    @field_validator("amenities")
    @classmethod
    def normalize_amenities(cls, values: list[str]) -> list[str]:
        return _normalize_amenities(values)


class ListingCreate(ListingBase):
    """Payload accepted when staff/admin creates a listing."""

    model_config = ConfigDict(from_attributes=True)


class ListingUpdate(BaseModel):
    """Partial listing update payload."""

    title: str | None = Field(None, min_length=1, max_length=150)
    status: ListingStatus | None = None
    is_public: bool | None = None
    is_featured: bool | None = None
    hide_exact_address: bool | None = None

    price: int | None = Field(None, ge=0, le=MAX_PRICE)
    property_type: PropertyType | None = None

    address: str | None = Field(None, min_length=1, max_length=255)
    city: str | None = Field(None, min_length=1, max_length=100)
    state: str | None = Field(None, min_length=2, max_length=2)

    description: str | None = Field(None, max_length=20_000)
    sqft: int | None = Field(None, ge=0, le=MAX_SQFT)
    acreage: float | None = Field(None, ge=0, le=MAX_ACREAGE)
    year_built: int | None = None

    bedrooms: int | None = Field(None, ge=0, le=MAX_BEDROOMS)
    bathrooms: float | None = Field(
        None,
        ge=0,
        le=MAX_BATHROOMS,
        multiple_of=0.5,
    )

    annual_property_taxes: float | None = Field(
        None,
        ge=0,
        le=MAX_MONEY_FIELD,
    )
    hoa_fee: float | None = Field(None, ge=0, le=MAX_MONEY_FIELD)
    hoa_fee_frequency: HoaFeeFrequency | None = None

    school_district: str | None = Field(None, max_length=150)
    amenities: list[str] | None = None

    mls_number: str | None = Field(None, max_length=100)
    source_attribution: str | None = Field(None, max_length=255)
    cover_image: str | None = Field(None, max_length=2048)

    @field_validator("state")
    @classmethod
    def normalize_state(cls, value: str | None) -> str | None:
        return _normalize_state(value) if value is not None else None

    @field_validator("year_built")
    @classmethod
    def validate_year_built(cls, value: int | None) -> int | None:
        return _validate_year_built(value)

    @field_validator("amenities")
    @classmethod
    def normalize_amenities(cls, values: list[str] | None) -> list[str] | None:
        return _normalize_amenities(values) if values is not None else None

    model_config = ConfigDict(from_attributes=True)


class ListingRead(ListingBase):
    """Listing returned by internal APIs."""

    id: int
    created_at: datetime | None = None
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class PaginatedListingRead(BaseModel):
    """Paginated internal listing response."""

    items: list[ListingRead]
    total: int
    page: int
    per_page: int

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/app/models/listing.ts
// Shared listing types used by the Angular admin application.

export const LISTING_STATUSES = [
  'Draft',
  'Active',
  'Pending',
  'Sold',
  'Archived',
] as const;

export type ListingStatus = (typeof LISTING_STATUSES)[number];

export const PUBLIC_LISTING_STATUSES: ListingStatus[] = [
  'Active',
  'Pending',
  'Sold',
];

export function isPubliclyEligibleStatus(status: ListingStatus): boolean {
  return PUBLIC_LISTING_STATUSES.includes(status);
}

export function effectiveListingVisibility(
  listing: Pick<ListingPayload, 'status' | 'is_public'>,
): 'Visible publicly' | 'Hidden from public' | 'Internal only' {
  if (!isPubliclyEligibleStatus(listing.status)) {
    return 'Internal only';
  }

  return listing.is_public ? 'Visible publicly' : 'Hidden from public';
}

export const PROPERTY_TYPES = [
  'Single Family',
  'Condo',
  'Townhouse',
  'Multi-Family',
  'Land',
  'Commercial',
  'Farm/Ranch',
  'Manufactured Home',
  'Other',
] as const;

export type PropertyType = (typeof PROPERTY_TYPES)[number];

export const HOA_FEE_FREQUENCIES = [
  'Monthly',
  'Quarterly',
  'Annually',
] as const;

export type HoaFeeFrequency = (typeof HOA_FEE_FREQUENCIES)[number];

export const MAX_LISTING_PRICE = 10_000_000_000;
export const MAX_LISTING_SQFT = 10_000_000;
export const MAX_LISTING_ACREAGE = 100_000_000;
export const MAX_LISTING_MONEY_FIELD = 100_000_000;
export const MAX_LISTING_BEDROOMS = 100;
export const MAX_LISTING_BATHROOMS = 100;

// Represents the editable listing data shared by create and update requests.
export interface ListingPayload {
  title: string;
  status: ListingStatus;
  is_public: boolean;
  is_featured: boolean;
  hide_exact_address: boolean;
  agent_id?: number | null;
  office_id?: number | null;

  price: number;
  property_type: PropertyType | null;

  address: string;
  city: string;
  state: string;

  description: string | null;
  sqft: number | null;
  acreage: number | null;
  year_built: number | null;

  bedrooms: number;
  bathrooms: number;

  annual_property_taxes: number | null;
  hoa_fee: number | null;
  hoa_fee_frequency: HoaFeeFrequency | null;

  school_district: string | null;
  amenities: string[];

  mls_number: string | null;
  source_attribution: string | null;
  cover_image: string | null;
  virtual_tour_url?: string | null;
}

// Represents one listing returned by the FastAPI backend.
export interface Listing extends ListingPayload {
  id: number;
  created_at: string | null;
  updated_at: string | null;
}

// Represents the paginated response returned by GET /listings.
export interface PaginatedListingsResponse {
  items: Listing[];
  total: number;
  page: number;
  per_page: number;
}

// Represents data sent to POST /listings.
export type ListingCreate = ListingPayload;

// Represents data sent to PUT /listings/{id}.
export type ListingUpdate = ListingPayload;


// Represents one staff-managed open-house schedule entry.
export interface OpenHouseEvent {
  id: number;
  listing_id: number;
  starts_at: string;
  ends_at: string;
  created_at: string;
  updated_at: string;
}

export interface OpenHousePayload {
  starts_at: string;
  ends_at: string;
}

export type OpenHouseCreate = OpenHousePayload;
export type OpenHouseUpdate = Partial<OpenHousePayload>;

export interface PublicOpenHouse {
  id: number;
  starts_at: string;
  ends_at: string;
}
